smurfFinalProject/smurfFinalProject/main.cpp

```diff
@@ -0,0 +1,198 @@
+
+auto grammar = R"(
+    program                 <-  code
+    code                    <-  statement*
+    function_call           <-  'print' '(' call_arguments ')' / variable_reference '(' call_arguments ')'
+    statement               <-  let_stmt / assignment / expr
+    let_stmt                <-  'let' variable_declaration
+    expr                    <-  boolean_expression / arithmetic_expression
+    variable_declaration    <-  decl (',' decl)*
+    decl                    <-  identifier ('=' expr)?
+    variable_reference      <-  identifier
+    assignment              <-  identifier '=' expr
+    boolean_expression      <-  arithmetic_expression rel_op arithmetic_expression
+    arithmetic_expression   <-  mult_term add_op arithmetic_expression / mult_term
+    mult_term               <-  primary mul_op mult_term / primary
+    primary                 <-  integer / function_call / variable_reference / '(' arithmetic_expression ')'
+    integer                 <-  < '-'? [0-9]+ >
+    add_op                  <-  < '+' / '-' >
+    mul_op                  <-  < '*' / '/' >
+    rel_op                  <-  < '==' / '!=' / '>=' / '>' / '<=' / '<' >
+    identifier              <-  < [a-z][a-zA-Z_0-9]* >
+    call_arguments          <-  (expr (',' expr)*)?
+    param_list              <-  '(' identifier (',' identifier)* ')' / '(' ')'
+    %whitespace             <-  [ \t\r\n]*
+)";
+
+//Grammar not added yet
+//expr                    <-  'if' if_expression / boolean_expression / arithmetic_expression
+//if_expression           <-  expr brace_block ( 'else' brace_block )?
+//brace_block             <-  '{' code '}'
+//function_definition     <-  param_list brace_block
+//comment                <-  '#' r'.*'
+
+class visitor;                                                      //The visitor pattern goes to each node to evaluate
+
+class ParseTreeNode
+{
+    node *content;
+public:
+    ParseTreeNode(){};
+    ParseTreeNode(node *content_node){
+        content = content_node;
+    }
+    node *get() const{
+        return content;
+    }
+    string to_string(){
+        return content->str();
+    }
+};
+
+node *bin_op(const SemanticValues &sv)                              //The binop function takes in the string passed to the parser
+{                                                                   //and breaks it into nodes and pointer nodes to create a smaller
+    node *left = sv[0].get<ParseTreeNode>().get();                  //tree within the larger AST consisting of two values as children
+    for (unsigned int i = 1; i < sv.size(); i += 2){                //and an operation node as the parent:
+        node *right = sv[i + 1].get<ParseTreeNode>().get();         //              operation
+        string op = sv[i].get<ParseTreeNode>().get()->str();        //              |       |
+        left = new binopNode(left, op, right);                      //          value       value
+    }
+    return left;
```

Comment:
Cool comments.  I used to do stuff like this when I was starting out.

As you continue to code, though, you'll discover that comments such as these become a burden: swap two lines in the code, and you have to adjust the comment. If you decide to split the function into two, you have to rewrite it.

Not a criticism: just a prediction :)